llvm/lib/Transforms/Scalar/GVN.cpp

```diff
@@ -2472,7 +2472,8 @@ bool GVNPass::propagateEquality(Value *LHS, Value *RHS,
       continue;
 
     // Prefer a constant on the right-hand side, or an Argument if no constants.
-    if (isa<Constant>(LHS) || (isa<Argument>(LHS) && !isa<Constant>(RHS)))
+    if (isa<Constant>(LHS) ||
+        (isa<Argument>(LHS) && !isa<Argument>(RHS) && !isa<Constant>(RHS)))
```

Comment:
What's the benefit of this change? Can you provide some motivating cases?

Reply:
I think this doesn't make a functional difference at all, because we will later sort the two Argument case by value number.

I think if you're going to refactor this, what you might want to do is extract the common code from the assume and equality handling, which both basically do the same thing in slightly different ways.